File: front/src/app/outfitter/schematic.ts
import {SchematicLayer} from "./schematic-layer";

export class Schematic {
  bodyType: string;
  bgColor?: string;
  bgPattern?: number;
  bodyScale?: string;
  layers: SchematicLayer[];

  constructor(bodyType: string) {
    this.bodyType = bodyType;
    this.layers = [];
  }

  static fromJSON(args: {
    bodyType: string,
    bgColor?: string,
    bgPattern?: number,
    bodyScale?: string,
    layers: {
      part: string,
      index: number,
      base?: string,
      detail?: string,
      outline?: string,
      opacity?: number,
      pattern?: number,
      shading?: number,
      resize?: number[],
      move?: number[],
      flip?: boolean
    }[]
  }): Schematic {
    let out = new Schematic(args.bodyType);
    if (args.bgColor) {
      out.bgColor = args.bgColor;
    }
    if (args.bgPattern) {
      out.bgPattern = args.bgPattern;
    }
    if (args.bodyScale) {
      out.bodyScale = args.bodyScale;
    }
    out.layers = args.layers.map((l) => SchematicLayer.fromJSON(l));
    return out;
  }

  toJSON(): {
    bodyType: string,
    bgColor?: string,
    bgPattern?: number,
    bodyScale?: string,
    layers: {
      part: string,
      index: number,
      base?: string,
      detail?: string,
      outline?: string,
      opacity?: number,
      pattern?: number,
      shading?: number,
      resize?: number[],
      move?: number[],
      flip?: boolean
    }[]
  } {
    let out: {
      bodyType: string,
      bgColor?: string,
      bgPattern?: number,
      bodyScale?: string,
      layers: {
        part: string,
        index: number,
        base?: string,
        detail?: string,
        outline?: string,
        opacity?: number,
        pattern?: number,
        shading?: number,
        resize?: number[],
        move?: number[],
        flip?: boolean
      }[]
    } = {
      bodyType: this.bodyType,
      layers: this.layers.map((layer) => layer.toJSON())
    };
    if (this.bgColor) {
      out['bgColor'] = this.bgColor;
    }
    if (this.bgPattern && this.bgPattern >= 0) {
      out['bgPattern'] = this.bgPattern;
    }
    if (this.bodyScale && this.bodyScale.length > 0) {
      out['bodyScale'] = this.bodyScale;
    }
    return out;
  }
}

File: src/carabiner/outfitter_old/schematic.clj
(ns carabiner.outfitter-old.schematic
  (:require [clojure.spec.alpha :as s]
            [carabiner.rogue94.common-schema :as rcs]
            [carabiner.outfitter-old.common-schema :as cs]
            [carabiner.outfitter-old.constants :as oc]
            [carabiner.common.regex :as r]
            [carabiner.common.schema :as ccs]))

(s/def ::body-types (set (mapv keyword oc/body-types)))

(s/def ::whole-number (s/and int? (complement neg?)))

(s/def ::percent
  (s/or :string (s/and string? (partial re-matches r/percent-pattern))
        :number (s/and number? ccs/is-double-percent?)))

(s/def ::layer (s/and
                 vector?
                 (s/cat :part-type ::cs/part-types
                        :index ::whole-number
                        :base (s/? (s/cat :label #{:base} :value ::rcs/color))
                        :detail (s/? (s/cat :label #{:detail} :value ::rcs/color))
                        :outline (s/? (s/cat :label #{:outline} :value ::rcs/color))
                        :opacity (s/? (s/cat :label #{:outline} :value ::percent))
                        :pattern (s/? (s/cat :label #{:pattern} :value ::whole-number))
                        :shading (s/? (s/cat :label #{:shading} :value ::whole-number))
                        :resize (s/? (s/cat :label #{:resize} :value ::cs/double-xy))
                        :move (s/? (s/cat :label #{:move} :value ::cs/double-xy))
                        :flip? (s/? (s/cat :label #{:flip?} :value boolean?)))))

(s/def ::bg-color ::rcs/color)

(s/def ::bg-pattern ::whole-number)

(s/def ::body-scale (set (mapv keyword oc/body-scale)))

(s/def ::schematic (s/and
                     vector?
                     (s/cat :body-type ::body-types
                            :args (s/? (s/keys :opt-un [::bg-color ::bg-pattern ::body-scale]))
                            :layers (s/+ ::layer))))
